tests/PHPStan/Reflection/ReflectionProviderGoldenTest.php

```diff
@@ -0,0 +1,663 @@
+				$visibility = 'public';
+				break;
+			case $methodReflection->isPrivate():
+				$visibility = 'private';
+				break;
+			default:
+				$visibility = 'protected';
+				break;
+		}
+
+		$result .= 'Visibility: ' . $visibility . "\n";
+		$result .= self::generateVariantsDescription($methodReflection->getName(), $methodReflection->getVariants());
+
+		return $result;
+	}
+
+	/** @param FunctionReflection|ExtendedMethodReflection $reflection */
+	private static function generateFunctionMethodBaseDescription($reflection): string
+	{
+		$result = '';
+
+		if (! $reflection->isDeprecated()->no()) {
+			$result .= 'Is deprecated: ' . $reflection->isDeprecated()->describe() . "\n";
+		}
+
+		if (! $reflection->isFinal()->no()) {
+			$result .= 'Is final: ' . $reflection->isFinal()->describe() . "\n";
+		}
+
+		if (! $reflection->isInternal()->no()) {
+			$result .= 'Is internal: ' . $reflection->isInternal()->describe() . "\n";
+		}
+
+		if (! $reflection->returnsByReference()->no()) {
+			$result .= 'Returns by reference: ' . $reflection->returnsByReference()->describe() . "\n";
+		}
+
+		if (! $reflection->hasSideEffects()->no()) {
+			$result .= 'Has side-effects: ' . $reflection->hasSideEffects()->describe() . "\n";
+		}
+
+		if ($reflection->getThrowType() !== null) {
+			$result .= 'Throw type: ' . $reflection->getThrowType()->describe(VerbosityLevel::precise()) . "\n";
+		}
+
+		return $result;
+	}
+
+	/** @param ParametersAcceptorWithPhpDocs[] $variants */
+	private static function generateVariantsDescription(string $name, array $variants): string
+	{
+		$variantCount = count($variants);
+		$result = 'Variants: ' . $variantCount . "\n";
+		$variantIdent = '    ';
+		$verbosityLevel = VerbosityLevel::precise();
+
+		foreach ($variants as $variant) {
+			$paramsNative = [];
+			$paramsPhpDoc = [];
+
+			foreach ($variant->getParameters() as $param) {
+				$paramsPhpDoc[] = $variantIdent . ' * @param ' . $param->getType()->describe($verbosityLevel) . ' $' . $param->getName() . "\n";
+
+				if ($param->getOutType() !== null) {
+					$paramsPhpDoc[] = $variantIdent . ' * @param-out ' . $param->getOutType()->describe($verbosityLevel) . ' $' . $param->getName() . "\n";
+				}
+
+				$passedByRef = $param->passedByReference();
+
+				if ($passedByRef->no()) {
+					$refDes = '';
+				} elseif ($passedByRef->createsNewVariable()) {
+					$refDes = '&rw';
+				} else {
+					$refDes = '&r';
+				}
+
+				$variadicDesc = $param->isVariadic() ? '...' : '';
+				$defValueDesc = $param->getDefaultValue() !== null
+					? ' = ' . $param->getDefaultValue()->describe($verbosityLevel)
+					: '';
+
+				$paramsNative[] = $param->getNativeType()->describe($verbosityLevel) . ' ' . $variadicDesc . $refDes . '$' . $param->getName() . $defValueDesc;
+			}
+
+			$result .= $variantIdent . "/**\n";
+			$result .= implode('', $paramsPhpDoc);
+			$result .= $variantIdent . ' * @return ' . $variant->getReturnType()->describe($verbosityLevel) . "\n";
```

Comment:
> Otherwise you're gonna generate invalid PHPDocs.

I don't think that's relevant. The "PHP" code is not valid either.

I want to have as much detail as possible in a human readable form. I tried using the Printer for return type, but some detail is lost. Here are some examples:

```
603) PHPStan\Reflection\ReflectionProviderGoldenTest::test with data set "METHOD WeakReference::get" ('METHOD WeakReference::get', 'Has side-effects: Maybe\nVisi...t|null')
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 Visibility: public
 Variants: 1
     /**
-     * @return T of object (class WeakReference, parameter)|null
+     * @return T|null
      */
     function get(): object|null'

598) PHPStan\Reflection\ReflectionProviderGoldenTest::test with data set "METHOD Vtiful\Kernel\Format::underline" ('METHOD Vtiful\Kernel\Format::underline', 'Is internal: Yes\nHas side-ef...Format')
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 Variants: 1
     /**
      * @param int $style
-     * @return static(Vtiful\Kernel\Format)
+     * @return static
      */
     function underline(int $style): Vtiful\Kernel\Format'

584) PHPStan\Reflection\ReflectionProviderGoldenTest::test with data set "METHOD SplTempFileObject::getType" ('METHOD SplTempFileObject::getType', 'Has side-effects: Maybe\nVisi... mixed')
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 Visibility: public
 Variants: 1
     /**
-     * @return (string|false)
+     * @return string|false
      */
     function getType(): mixed'

582) PHPStan\Reflection\ReflectionProviderGoldenTest::test with data set "METHOD SplTempFileObject::getRealPath" ('METHOD SplTempFileObject::getRealPath', 'Has side-effects: Maybe\nVisi... mixed')
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 Visibility: public
 Variants: 1
     /**
-     * @return (string|false)
+     * @return string|false
      */
     function getRealPath(): mixed'
```

Reply:
Alright, understood.